Sources/PackageGraph/PinsStore.swift

```diff
@@ -84,93 +72,186 @@ public final class PinsStore {
     /// - Parameters:
     ///   - packageRef: The package reference to pin.
     ///   - state: The state to pin at.
-    public func pin(
-        packageRef: PackageReference,
-        state: CheckoutState
-    ) {
-        self.pinsMap[packageRef.identity] = Pin(
+    public func pin(packageRef: PackageReference, state: CheckoutState) {
+        self.add(.init(
             packageRef: packageRef,
             state: state
-        )
+        ))
     }
 
     /// Add a pin.
     ///
     /// This will replace any previous pin with same package name.
     public func add(_ pin: Pin) {
-        self.pinsMap[pin.packageRef.identity] = pin
+        self._pins[pin.packageRef.identity] = pin
     }
 
     /// Unpin all of the currently pinned dependencies.
     ///
     /// This method does not automatically write to state file.
     public func unpinAll() {
         // Reset the pins map.
-        self.pinsMap = [:]
+        self._pins.clear()
     }
 
     public func saveState() throws {
-        if self.pinsMap.isEmpty {
+        try self.storage.save(pins: self._pins.get(), mirrors: self.mirrors, removeIfEmpty: true)
+    }
+}
+
+// MARK: - Serialization
+
+fileprivate struct PinsStorage {
+    private let path: AbsolutePath
+    private let fileSystem: FileSystem
+    private let encoder = JSONEncoder.makeWithDefaults()
+    private let decoder = JSONDecoder.makeWithDefaults()
+
+    init(path: AbsolutePath, fileSystem: FileSystem) {
+        self.path = path
+        self.fileSystem = fileSystem
+    }
+
+    func load(mirrors: DependencyMirrors) throws -> PinsStore.PinsMap {
+        if !self.fileSystem.exists(self.path) {
+            return [:]
+        }
+
+        // 👀 do we want a lock file here? safer but pretty noisy
```

Comment:
I think so, if there's any risk of, say, Xcode and swift-package CLI both accessing at the same time.

Reply:
@abertelrud it would mean we will add a `Package.resolved.lock` file which is safe but noisy, also I think Xcode has its own location for `Package.resolved` files?

@neonichu wdyt?

Reply:
Ok, so this is indeed `Package.resolved` (I wasn't sure if the pin store was the same thing).  I would think that the `.lock` goes next to wherever the IDE says to put `Package.resolved` — in the case of Xcode, that's down in one of its semihidden support directories, so it might not be particularly visible.  It feels unsafe not to lock it.